### PR TITLE
Golang library for validation: folder structure validation

### PR DESCRIPTION
**Pull request by ycombinator**
This PR implements validation of a package's folder structure. Structure of package files is not yet implemented and will be done in a follow up PR.

Related: #4.

### CODE REVIEW

**Comment by ycombinator** [comment]
@ruflin @mtojek This PR is a WIP but using it in its current state to validate all packages in the @elastic/integrations repo produces the following result. I suspect some of the following validation errors need fixes in the package while others need fixes in the spec. Could you please let me know which is which. Thank you!

```
Validating package [apache]...
Error: found 7 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/access/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/access/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/access]
   4. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/error/agent/stream]
   5. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/error/agent/stream]
   6. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/error]
   7. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/apache]

Validating package [aws]...
Error: found 21 validation errors:
   1. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudtrail/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudtrail/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudtrail]
   4. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudwatch_logs/agent/stream]
   5. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudwatch_logs/agent/stream]
   6. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/cloudwatch_logs]
   7. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/ec2_logs/agent/stream]
   8. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/ec2_logs/agent/stream]
   9. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/ec2_logs]
  10. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/elb_logs/agent/stream]
  11. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/elb_logs/agent/stream]
  12. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/elb_logs]
  13. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/s3access/agent/stream]
  14. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/s3access/agent/stream]
  15. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/s3access/agent/stream]
  16. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/s3access]
  17. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/vpcflow/agent/stream]
  18. item [s3.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/vpcflow/agent/stream]
  19. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/vpcflow/agent/stream]
  20. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/aws/dataset/vpcflow]
  21. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/aws]

Validating package [cisco]...
Error: found 7 validation errors:
   1. item [udp.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/asa/agent/stream]
   2. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/asa]
   3. item [udp.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/ftd/agent/stream]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/ftd]
   5. item [syslog.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/ios/agent/stream]
   6. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/cisco/dataset/ios]
   7. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/cisco]

Validating package [crowdstrike]...
Error: found 5 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon/agent/stream]
   3. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon]
   5. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike]

Validating package [haproxy]...
Error: found 5 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/haproxy/dataset/log/agent/stream]
   2. item [syslog.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/haproxy/dataset/log/agent/stream]
   3. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/haproxy/dataset/log/agent/stream]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/haproxy/dataset/log]
   5. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/haproxy]

Validating package [iis]...
Error: found 5 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/iis/dataset/access/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/iis/dataset/access/agent/stream]
   3. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/iis/dataset/error/agent/stream]
   4. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/iis/dataset/error/agent/stream]
   5. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/iis]

Validating package [kafka]...
Error: found 4 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/kafka/dataset/log/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/kafka/dataset/log/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/kafka/dataset/log]
   4. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/kafka]

Validating package [kubernetes]...
Error: found 1 validation error:
   1. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/kubernetes]

Validating package [log]...
Error: found 2 validation errors:
   1. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/log/dataset/log]
   2. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/log]

Validating package [mongodb]...
Error: found 3 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/mongodb/dataset/log/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/mongodb/dataset/log/agent/stream]
   3. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/mongodb]

Validating package [mysql]...
Error: found 5 validation errors:
   1. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/error]
   2. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]
   3. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog]
   5. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/mysql]

Validating package [netflow]...
Error: found 4 validation errors:
   1. item [netflow.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/netflow/dataset/log/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/netflow/dataset/log/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/netflow/dataset/log]
   4. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/netflow]

Validating package [nginx]...
Error: found 5 validation errors:
   1. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/access]
   2. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/access]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/error]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/ingress_controller]
   5. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/nginx]

Validating package [postgresql]...
Error: found 4 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/postgresql/dataset/log/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/postgresql/dataset/log/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/postgresql/dataset/log]
   4. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/postgresql]

Validating package [prometheus]...
Error: found 1 validation error:
   1. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/prometheus]

Validating package [rabbitmq]...
Error: found 4 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/rabbitmq/dataset/log/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/rabbitmq/dataset/log/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/rabbitmq/dataset/log]
   4. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/rabbitmq]

Validating package [redis]...
Error: found 7 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   2. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log]
   4. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]
   5. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]
   6. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog]
   7. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis]

Validating package [suricata]...
Error: found 4 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/suricata/dataset/eve/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/suricata/dataset/eve/agent/stream]
   3. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/suricata/dataset/eve]
   4. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/suricata]

Validating package [system]...
Error: found 27 validation errors:
   1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth/agent/stream]
   3. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth/elasticsearch/ingest_pipeline]
   4. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth]
   5. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/core]
   6. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/cpu]
   7. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/diskio]
   8. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/entropy]
   9. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/filesystem]
  10. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/fsstat]
  11. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/load]
  12. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/memory]
  13. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/network]
  14. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/network_summary]
  15. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/process]
  16. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/process_summary]
  17. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/raid]
  18. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/service]
  19. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/socket]
  20. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/socket_summary]
  21. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog/agent/stream]
  22. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog/agent/stream]
  23. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog/elasticsearch/ingest_pipeline]
  24. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog]
  25. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/uptime]
  26. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/users]
  27. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/system]

Validating package [windows]...
Error: found 3 validation errors:
   1. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/windows/dataset/perfmon]
   2. expecting to find [sample_event.json] file in folder [/Users/shaunak/development/github/integrations/packages/windows/dataset/service]
   3. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/windows]

Validating package [zookeeper]...
Error: found 1 validation error:
   1. expecting to find [changelog.yml] file in folder [/Users/shaunak/development/github/integrations/packages/zookeeper]
```

**Comment by mtojek** [comment]
Let me refer to errors. I'm afraid all of them slipped through the review phase.

1. Item is not allowed in folder

```
1. item [log.yml.hbs] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/access/agent/stream]
   2. expecting to find [stream.yml.hbs] file in folder [/Users/shaunak/development/github/integrations/packages/apache/dataset/access/agent/stream]
```

I think this is a valid file. Take a look at the `manifest.yml` file for the given dataset. It has a reference to it (`template_path: log.yml.hbs`). 

2. Missing `sample_event.json` file

```
expecting to find [sample_event.json] file in folder
```

This is optional at the moment. You can use this file to render samples in docs. I'm not convinces whether it should be mandatory or not.

3. Expecting to find [changelog.yml] file in folder

I'm allergic to changelogs, probably @ruflin would like to elaborate on them :) I suppose we will have to start using them, but I'm not sure if people get mad, when you ask them to fill them up.

File: code/go/internal/validator/folder.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	"path"
+	"path/filepath"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+)
+
+const ItemTypeFile = "file"
+const ItemTypeFolder = "folder"
```
Comment (mtojek) [inline code/go/internal/validator/folder.go:17]: should theses consts be exposed?

File: code/go/pkg/validator/validator.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package validator
+
+import (
+	"github.com/elastic/package-spec/code/go/internal/validator"
+)
+
+// ValidateFromPath validates a package located at the given path against the
+// appropriate specification and returns any errors.
+func ValidateFromPath(packageRootPath string) error {
```
Comment (mtojek) [inline code/go/pkg/validator/validator.go:9]: This is the method I was referring in different PR: ValidateFromPath. I would be cool not to load specs every time.

File: code/go/pkg/validator/validator.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	pkg, err := validator.NewPackage(packageRootPath)
+	if err != nil {
+		return err
+	}
+
+	spec, err := validator.NewSpec(*pkg.SpecVersion)
```
Comment (mtojek) [inline code/go/pkg/validator/validator.go:15]: Is there a chance that `pkg.SpecVersion` is nil?

File: code/go/internal/validator/spec.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package validator
+
+import (
+	"net/http"
+	"path"
+	"strconv"
+
+	"github.com/Masterminds/semver/v3"
+	_ "github.com/elastic/package-spec/code/go/internal/spec"
```
Comment (mtojek) [inline code/go/internal/validator/spec.go:9]: nit: sort imports

File: code/go/pkg/validator/validator.go (hunk 2 above)
Comment (ycombinator) [reply code/go/pkg/validator/validator.go:9]: Yep, see my reply on the other PR: https://github.com/elastic/package-spec/pull/12#discussion_r467763167. Let's continue the conversation there.

**Comment by ruflin** [comment]
@mtojek For changelog: We need a way to tell users of the packages on what change between versions. I'm not tied at all to changelogs but we need a solution.

**Comment by ycombinator** [comment]
My 2 cents: I like the proposed CHANGELOG yml format. It captures what we need, but in a structured fashion. So one can imagine generating human-readable changelog documentation from it, etc.

> I'm allergic to changelogs,

@mtojek could you elaborate on why this is? maybe there's a drawback that I'm not seeing currently.

File: code/go/internal/validator/spec.go (hunk 4 above)
Comment (ycombinator) [reply code/go/internal/validator/spec.go]: Fixed in 7741495a40da3ff61c5ab3c35bbcd61014e3f1a6

File: code/go/internal/validator/folder.go (hunk 1 above)
Comment (ycombinator) [reply code/go/internal/validator/folder.go]: Fixed in 11f07afc3013fc66a81701d2e64fde29de87900a.

File: code/go/pkg/validator/validator.go (hunk 3 above)
Comment (ycombinator) [reply code/go/pkg/validator/validator.go:21]: It should not happen but I added a guard, just in case, in 569b89b.

**Comment by urso** [comment]
> My 2 cents: I like the proposed CHANGELOG yml format. It captures what we need, but in a structured fashion. So one can imagine generating human-readable changelog documentation from it, etc.
>> I'm allergic to changelogs,

And I'm getting tired of yml :)

What issue does using YAML solve here?

**Comment by ycombinator** [comment]
I'm not married to the YML but I like that it provides some structure. I suppose we would get enough structure from using asciidoc too, as we did in Beats, or some other format. So I'm good with either. But I do think it's useful to have package CHANGELOGs so we can communicate to users what changed in a package version.

**Comment by mtojek** [comment]
> @mtojek could you elaborate on why this is? maybe there's a drawback that I'm not seeing currently.

The same one as for Beats and Package-Registry - it's really easy to forget about it and make it inconsistent. I'm thinking about generating it out of commit messages. At least we'll have an order in Git log :)

**Comment by ycombinator** [comment]
Alright, sounds like we don't have consensus on how the CHANGELOG will be generated for packages and what format it might be stored in, if it is stored at all. So, for the scope of this PR here, I'm going to remove `changelog.yml` as a required file in the package root folder. Let's discuss CHANGELOG handling for packages separately and once we have consensus we can add validation for it, if appropriate, in the right place.

**Comment by urso** [comment]
> it's really easy to forget about it and make it inconsistent. I'm thinking about generating it out of commit messages. At least we'll have an order in Git log :)

For Beats we want to move to adding Changelog information to the PR description. Changelog are a pain and often need editing.

> I'm not married to the YML but I like that it provides some structure while also being pretty much human-readable

I wonder how much value the format provides. Using YAML means we have to convert/process the file eventually. What is the structure actually required for (maybe I'm missing something)?

The advantage of YAML is that it is easier to process for automation purposes.

With structure you still have to be careful. One reason we added CHANGELOG.next.asciidoc to Beats is because developers are often wrong about the actual version a change is include. If releases get postponed the changelog can actually be wrong again (yep, maintaining Beats changelog was a not so fun task). I guess it is not so much a problem for integrations, but putting the release version into your YAML structure per entry is potentially asking for some heavy editing later on (also depends on FF and release procedures and policies).

All in all it would be nice if we eventually settle for a common way on how to handle changelogs for related projects, in order to make live easier for developers that have to touch all these related projects.

**Comment by ycombinator** [comment]
I've created https://github.com/elastic/package-spec/issues/17 so we can have a discussion dedicated to changelogs over there. I'll copy over some of the comments from here so they are readily available.

**Comment by mtojek** [comment]
I'm fine with taking the CHANGELOG discussion out the scope of this PR. I believe you can mark it as optional or remove its definition from spec.

**Comment by mtojek** [comment]
CI complains about missing comments:

```
[2020-08-11T22:30:57.434Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/package.go:14:6: exported type Package should have comment or be unexported
[2020-08-11T22:30:57.434Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/package.go:19:1: exported function NewPackage should have comment or be unexported
[2020-08-11T22:30:57.434Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/spec.go:12:2: a blank import should be only in a main or test package, or have a comment justifying it
[2020-08-11T22:30:57.434Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/spec.go:15:6: exported type Spec should have comment or be unexported
[2020-08-11T22:30:57.435Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/spec.go:21:1: exported function NewSpec should have comment or be unexported
[2020-08-11T22:30:57.435Z] /var/lib/jenkins/workspace/Beats_package-spec_PR-14/src/github.com/elastic/package-spec/code/go/internal/validator/spec.go:43:1: exported method Spec.ValidatePackage should have comment or be unexported
```

**Comment by ruflin** [comment]
++ on skipping changelog and follow up.

**Comment by ycombinator** [comment]
After making updates to the folder validation per review feedback, validating all existing packages in the @elastic/integrations repository with this PR produces the following results.

```
Validating package [apache]...
Validating package [aws]...
Validating package [checkpoint]...
Validating package [cisco]...
Validating package [crowdstrike]...
Error: found 1 validation error:
   1. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon]

Validating package [fortinet]...
Error: found 3 validation errors:
   1. item [event.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]
   2. item [traffic.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]
   3. item [utm.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]

Validating package [haproxy]...
Validating package [iis]...
Validating package [kafka]...
Validating package [kubernetes]...
Validating package [log]...
Validating package [microsoft]...
Validating package [mongodb]...
Validating package [mysql]...
Error: found 2 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]
   2. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]

Validating package [netflow]...
Validating package [nginx]...
Error: found 1 validation error:
   1. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/access]

Validating package [o365]...
Validating package [okta]...
Validating package [panw]...
Validating package [postgresql]...
Validating package [prometheus]...
Validating package [rabbitmq]...
Validating package [redis]...
Error: found 4 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   2. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   3. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]
   4. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]

Validating package [suricata]...
Validating package [system]...
Error: found 2 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth/elasticsearch/ingest_pipeline]
   2. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog/elasticsearch/ingest_pipeline]

Validating package [windows]...
Validating package [zookeeper]...
```

As you can see most packages are now valid, which is nice! Let's go through the remaining errors reported above.

```
Validating package [crowdstrike]...
Error: found 1 validation error:
   1. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/crowdstrike/dataset/falcon]
...
Validating package [nginx]...
Error: found 1 validation error:
   1. item [test] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/nginx/dataset/access]
```

I think we have to allow this for now. There is a discussion I have started (#22) that might impact this validation but for now, I think we add it. I will update this PR to do so but just wanted to call it out here first in case anyone has any objections.

 ```
Validating package [fortinet]...
Error: found 3 validation errors:
   1. item [event.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]
   2. item [traffic.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]
   3. item [utm.yml] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline]
```

I checked and the `fortinet` package's `firewall` dataset [does define a `default.yml` ingest pipeline](https://github.com/elastic/integrations/blob/master/packages/fortinet/dataset/firewall/elasticsearch/ingest_pipeline/default.yml). This pipeline references the other three pipelines shown above. So I believe it's okay to have these files in there **as long as Kibana knows to load these additional ingest pipelines into Elasticsearch.** @ruflin can you confirm this, please?

```
Validating package [mysql]...
Error: found 2 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]
   2. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/mysql/dataset/slowlog/elasticsearch/ingest_pipeline]
...
Validating package [redis]...
Error: found 4 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   2. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/log/elasticsearch/ingest_pipeline]
   3. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]
   4. expecting to find [default.yml] file in folder [/Users/shaunak/development/github/integrations/packages/redis/dataset/slowlog/elasticsearch/ingest_pipeline]
...
Validating package [system]...
Error: found 2 validation errors:
   1. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/auth/elasticsearch/ingest_pipeline]
   2. item [default.json] is not allowed in folder [/Users/shaunak/development/github/integrations/packages/system/dataset/syslog/elasticsearch/ingest_pipeline]
```

So these errors look like a case of ingest pipelines being defined as JSON instead of YAML. What do we want to do here? I know Beats could accept either format — is this what we want for Packages too? or do we want to require that packages only define their ingest pipelines in YAML (in which case the above ingest pipelines will need to be converted to YAML)? 

Also, I assume Kibana is able to load ingest pipelines from either format into Elasticsearch? @ruflin?

**Comment by ruflin** [comment]
Kibana knows how to load multiple ingest pipelines. @skh @neptunian to make sure this is correct. 

For json vs yaml: I think we should support both. Yaml allows us when we build ingest pipelines from scratch to add comments etc, but users just exporting pipelines from Elasticsearch and copy / paste it to a package will likely use json. Kibana supports both formats. (@skh )

**Comment by neptunian** [comment]
> Kibana knows how to load multiple ingest pipelines. @skh @neptunian to make sure this is correct.

Yes.  The ingest manager will install all pipelines in the `ingest_pipeline` directory, not just the default or entry pipeline.


> For json vs yaml: I think we should support both. Yaml allows us when we build ingest pipelines from scratch to add comments etc, but users just exporting pipelines from Elasticsearch and copy / paste it to a package will likely use json. Kibana supports both formats. 

Yes, both formats are supported.

**Comment by ycombinator** [comment]
Thanks @ruflin and @neptunian for weighing in. Based on your feedback, I have now updated this PR. And as a result all integrations packages are now passing validation on this PR!

```
Validating package [apache]...
Validating package [aws]...
Validating package [checkpoint]...
Validating package [cisco]...
Validating package [crowdstrike]...
Validating package [fortinet]...
Validating package [haproxy]...
Validating package [iis]...
Validating package [kafka]...
Validating package [kubernetes]...
Validating package [log]...
Validating package [microsoft]...
Validating package [mongodb]...
Validating package [mysql]...
Validating package [netflow]...
Validating package [nginx]...
Validating package [o365]...
Validating package [okta]...
Validating package [panw]...
Validating package [postgresql]...
Validating package [prometheus]...
Validating package [rabbitmq]...
Validating package [redis]...
Validating package [suricata]...
Validating package [system]...
Validating package [windows]...
Validating package [zookeeper]...
```

@mtojek @ruflin would you mind re-reviewing this PR please? Thanks!

**Pull request merged**